### DISCUSSION TITLE
Get Vertices of b2PolygonShape?

### DISCUSSION THREAD

**Discussion by SammyJoeOsborne** [Birch-san/box2d-wasm#20] [Q&A] [answered]
Hello! I'm creating my own renderer, but discovering that the GetVertexCount() and GetVertex() methods seem to be missing from b2PolygonShape. Is there another approach or something I'm missing?

**Comment by Birch-san** [comment, marked as answer]
`b2PolygonShape#get_m_count()` and `b2PolygonShape#get_m_vertices(index?: number)`.

https://github.com/Birch-san/box2d-wasm/blob/master/docs/iteration.md

```js
const square = new b2PolygonShape();
square.SetAsBox(1, 1);
for (let vertexIx = 0; vertexIx < square.get_m_count(); vertexIx++) {
  const vertex = square.get_m_vertices(vertexIx);
}
```

**Reply by SammyJoeOsborne** [reply]
There seems to be a problem with the shape once the fixture is created. If you iterate through the world bodies, then iterate through the body's fixtures, then get the shape from the fixture, the shape object no longer has the get_m_count() or get_m_vertices() methods. Even though GetType still reports as "2" (polygon), the only available methods are get_m_radius() and get_m_type()

I'm going to try to put together a jsfiddle to demonstrate, but in the meantime you can verify by creating a world and creating this single body in it, then check the properties of firstShape. You will see it does not contain the get_m_count or get_m_vertices methods:

```
const square = new b2PolygonShape();
square.SetAsBox(1, 1);

const zero = new b2Vec2(0, 0);

const bd = new b2BodyDef();
bd.set_type(b2_dynamicBody);
bd.set_position(zero);

const body = world.CreateBody(bd);
body.CreateFixture(square, 1);

const firstBody = world.GetBodyList();
const firstFixture = firstBody.GetFixtureList();
const firstShape = firstFixture.GetShape();
firstShape.get_m_count(); // <-- throws an error. Inspect it and only get_m_radius and get_m_type exist
```

**Reply by SammyJoeOsborne** [reply]
JS Fiddle replicating issue here:
https://jsfiddle.net/slopps/jxdy2vck/13/

**Reply by Birch-san** [reply]
[`b2_shape`](https://github.com/erincatto/box2d/blob/master/include/box2d/b2_shape.h#L50) doesn't have those methods.

use `castObject` to cast it to the `b2PolygonShape` subclass.

```js
/** @type {Set<number>} */
const polygonShapes = new Set();

//…

const square = new b2PolygonShape();
square.SetAsBox(1, 1);
polygonShapes.add(getPointer(square));

//…

if (polygonShapes.has(getPointer(firstShape))) {
  const polygonShape = castObject(firstShape, b2PolygonShape);
  for (let vertexIx = 0; vertexIx < square.get_m_count(); vertexIx++) {
    const vertex = square.get_m_vertices(vertexIx);
  }
}
```

**Reply by SammyJoeOsborne** [reply]
Very kind of you to reply. I was hoping there might be a way to cast. I'll give this a shot and let you know how it turns out. Thanks again!

**Comment by SammyJoeOsborne** [comment]
The casting worked with some modification and I'm successfully rendering shapes by looping through the world bodies -> fixtures -> shape -> vertices. I really appreciate the help!

In case anyone sees these posts in the future, just want to clarify that the suggested polygonShapes Set lookup approach won't work. If you retrieve a shape from a fixture via b2Fixture.GetShape(), the retrieved shape does not have the same pointer as the original shape used to create the fixture. However, you can still cast the retrieved shape from the fixture and it will have the correct vertices.

You can cast it on an as-needed basis as suggested, but if its in a processor intensive loop this may be too cumbersome. I opted to attach a casted shape to every fixture I created so that I could reference the "typedShape" in intensive render loops:
```
createBox(x, y, width, height, angle = 0)
{
    const boxBodyDef = new b2BodyDef();
    boxBodyDef.angle = angle;
    boxBodyDef.position.Set(x, y);

    const boxShape = new b2PolygonShape();
    boxShape.SetAsBox(width / 2, height / 2);

    const boxBody = this.world.CreateBody(boxBodyDef);
    const fixture = boxBody.CreateFixture(boxShape, 0.0);
    fixture.typedShape = castObject(fixture.GetShape(), b2PolygonShape);

    return boxBody;
}
```
